distage/distage-testkit-scalatest/src/test/scala/izumi/distage/testkit/distagesuite/DistageTestExample.scala

```diff
@@ -37,15 +40,66 @@ class DistageTestExampleBIO extends DistageBIOSpecScalatest[zio.IO] with Distage
 
 }
 
-abstract class DistageTestExampleBase[F[_]: TagK](implicit F: DIEffect[F]) extends DistageSpecScalatest[F] with DistageMemoizeExample[F] {
+object DistageTestExampleBase {
+  sealed trait SetElement {
+    println(s"Here is $v element.")
+    def v: Int
+  }
+  final case class SetElement1(v: Int) extends SetElement
+  final case class SetElement2(v: Int) extends SetElement
+  final case class SetElement3(v: Int) extends SetElement
+}
+
+abstract class DistageTestExampleBase[F[_] : TagK](implicit F: DIEffect[F]) extends DistageSpecScalatest[F] with DistageMemoizeExample[F] {
 
   override protected def config: TestConfig = super.config.copy(
-      pluginSource = super.config.pluginSource ++ PluginSource(PluginConfig.cached(Seq("xxx")))
-    )
+    pluginSource = super.config.pluginSource ++ PluginSource(PluginConfig.cached(Seq("xxx"))) ++ new ModuleDef {
+      make[SetElement1].from(SetElement1(1))
+      make[SetElement2].from(SetElement2(2))
+      make[SetElement3].from(SetElement3(3))
+
+      many[SetElement]
+        .weak[SetElement1]
+        .weak[SetElement2]
+        .weak[SetElement3]
+
+      many[SetElement].named("set-id")
+        .weak[SetElement1]
+        .weak[SetElement2]
+        .weak[SetElement3]
+    }
+  )
 
   val XXX_Whitebox_memoizedMockCache = new AtomicReference[MockCache[F]]
 
   "distage test custom runner" should {
+    "support memoized weak sets" in {
+      (
+        set: Set[SetElement],
+        s1: SetElement1,
+        s2: SetElement2,
+        s3: SetElement3,
+      ) =>
+        Quirks.discard(s1, s2, s3)
+        F.maybeSuspend(assert(set.size == 3))
+    }
+
+    "support memoized named weak sets" in {
+      (
+        set: Set[SetElement]@Id("set-id"),
+        s1: SetElement1,
+        s2: SetElement2,
+        s3: SetElement3,
+      ) =>
+        Quirks.discard(s1, s2, s3)
+        F.maybeSuspend(assert(set.size == 3))
+    }
+
+    "return memoized weak set with have whole list of members even if test does not depends on them" in {
+      set: Set[SetElement] =>
```

Comment:
@neko-kai does we have any chance to skip this test if it runs without other? Without registry singleton using ?

Reply:
@Caparow 
You could use `assume` instead of `assert`, then the test would skip instead of failing when the predicate fails.

But, yeah, the test will fail if it's run individually, instead of with all tests:

```scala
Test failed
ScalaTestFailureLocation: izumi.distage.testkit.distagesuite.DistageTestExampleBase at (DistageTestExample.scala:100)
org.scalatest.exceptions.TestFailedException: Set() had size 0 instead of expected size 3
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
```

I think instead of skipping this test, we may probably want to always plan using _all_ tests in a test env and only filter them out later, after we've strengthened the weak components...

/cc @pshirshov

Reply:
@Caparow 
So, by @pshirshov , this behavior is as expected... You may add a constructions counter for `SetElement` and use this to skip test with `assume` or just `assume(set.nonEmpty)`...